optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
-== Filter
+=== Filter
+
+Filtering allows you to reduce the number of constraint matches in your stream.
+It first enumerate all constraint matches and then applies a predicate to filter some of them out.
+The predicate is a function that only returns `true` if the match is to continue in the stream.
+The following constraint stream will remove all of Beth's shifts from all `Shift` matches:
+
+[source,java,options="nowrap"]
+----
+    private Constraint penalizeAnnShifts(ConstraintFactory factory) {
+        return factory.from(Shift.class)
+                .filter(shift -> shift.getEmployeeName().equals("Ann"))
+                .penalize("Ann's shift", SimpleScore.ONE);
+    }
+----
+
+For a bi constraint match of `Shift` and `DayOff`, here's how to get a list of shifts where an employee has asked for a
+day off:
+
+[source,java,options="nowrap"]
+----
+    private Constraint penalizeShiftsOnOffDays(ConstraintFactory factory) {
+        return factory.from(Shift.class)
+                .join(DayOff.class)
+                .filter((shift, dayOff) -> shift.date == dayOff.date && shift.employee == dayOff.employee)
+                .penalize("Shift on an off-day", SimpleScore.ONE);
+    }
+----
+
+The following figure illustrates both these examples:
 
 image::ConstraintStreams/constraintStreamFilter.png[align="center"]
 
+[NOTE]
+====
+For reasons of performance, using the <<constraintStreamsJoin,join>> building block with appropriate `Joiner`s is
+preferrable when possible.
+Using a `Joiner` creates only the constraint matches that are necessary, while filtered join creates all possible
+constraint matches and only then filters some of them out.
+====
+
+For reference, we include a list of predicate functions required for filtering constraint streams of different
+cardinality:
+
+|===
+|Stream Cardinality|Filtering Predicate
+|1                 |`java.util.function.Predicate<A>`
+|2                 |`java.util.function.BiPredicate<A, B>`
+|3                 |`org.optaplanner.core.api.function.TriPredicate<A, B, C>`
+|4                 |`org.optaplanner.core.api.function.QuadPredicate<A, B, C, D>`
+|===
+
 
 [[constraintStreamsJoin]]
-== Join
+=== Join
+
+Joining is a way to increase <<constraintStreamsCardinality,stream cardinality>> and it is similar to inner join
+operation in SQL. As the following figure illustrates, a join creates a cartesian product of the streams being joined:
 
 image::ConstraintStreams/constraintStreamJoinWithoutJoiners.png[align="center"]
 
+This is inefficient, as the resulting stream may contain constraint matches that are of no interest to your constraint.
+Therefore, we recommend using `Joiner` to restrict your joins only to the matches you are actually interested in, as
+illustrated in this figure:
+
 image::ConstraintStreams/constraintStreamJoinWithJoiners.png[align="center"]
 
+For a list of all supported `Joiner` implementations, please refer to the
+`org.optaplanner.core.api.score.stream.Joiners` class.
```

Comment:
We should list all of them here (but not their overloaded methods), so something like:
- equalTo
- greaterThan
- ...